Item controller 1

cdpiazza opened the pull request:
1st Iteration

Commit:
En la rama item-controller-1
 Tu rama está actualizada con 'origin/item-controller-1'.
 Cambios a ser confirmados:
	nuevo archivo:  src/main/java/com/app/rest/controller/ItemController.java
	nuevo archivo:  src/main/java/com/app/rest/model/dto/ItemDTO.java
	nuevo archivo:  src/main/java/com/app/rest/model/dto/ItemType.java
	modificado:     src/main/java/com/app/rest/service/ItemService.java
	nuevo archivo:  src/test/java/com/app/rest/test/service/ItemTest.java

Commit:
En la rama item-controller-1
 Tu rama está actualizada con 'origin/item-controller-1'.
 Cambios a ser confirmados:
	modificado:     src/main/java/com/app/rest/controller/ItemController.java
	modificado:     src/main/java/com/app/rest/service/ItemService.java

Commit:
En la rama item-controller-1
 Tu rama está actualizada con 'origin/item-controller-1'.
 Cambios a ser confirmados:
	modificado:     src/main/java/com/app/rest/controller/ItemController.java
	modificado:     src/main/java/com/app/rest/model/persistence/ItemDetail.java
	modificado:     src/main/java/com/app/rest/service/ItemService.java

Commit:
En la rama item-controller-1
 Tu rama está adelantada a 'origin/item-controller-1' por 1 commit.
   (usa "git push" para publicar tus commits locales)
 Cambios a ser confirmados:
	modificado:     src/main/java/com/app/rest/controller/ItemController.java
	modificado:     src/main/java/com/app/rest/model/dto/ItemDTO.java
	modificado:     src/main/java/com/app/rest/service/ItemService.java
	borrado:        src/test/java/com/app/rest/test/service/DatabaseConnTest.java
	modificado:     src/test/java/com/app/rest/test/service/ItemTest.java

Commit:
Cambios a ser confirmados:
	modificado:     src/main/java/com/app/rest/controller/ItemController.java

alangarciac commented on src/main/java/com/app/rest/controller/ItemController.java:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    @PostMapping("/items/new")
+    public ResponseEntity<?> createItem(@RequestBody ItemDTO itemDTO) {
```
podríamos devolver el id del Item creado `ResponseEntity<Long>`

cdpiazza replied:
Fijate q lo devuelve, el tema es que si no lo pongo asi y lo pongo como ResponseEntity<ItemDTO>, hay un problema de tipo de la responseentity de la excepcion, no me deja cambiarle el message y hacer el build (algo no le gusta de los tipos), pero si, cuando anda todo se devuelve el itemDTO.

alangarciac replied:
perfecto entonces

alangarciac commented on src/main/java/com/app/rest/controller/ItemController.java:
```diff
@@ -0,0 +1,82 @@
+
+    }
+    @DeleteMapping("/items/{id}")
+    public ResponseEntity<?> deleteItem(@PathVariable("id") Long id) {
```
lo mismo acá `ResponseEntity<Long>` con el id del Item borrado

cdpiazza replied:
aca devuelvo el itemDTO completo, y esta con <?> por el mismo motivo q el create, para poder modificar el codigo de error del response entity en la exception.

alangarciac replied:
perfecto

alangarciac commented on src/main/java/com/app/rest/controller/ItemController.java:
```diff
@@ -0,0 +1,82 @@
+    @PutMapping("/items/{id}")
+    public ResponseEntity<?> updateItem(@PathVariable("id") Long id, @RequestBody ItemDTO itemDTO) {
+        try {
+            itemDTO.validate();   //Debo validar porque el DTO se autocarga con el json usando reflexion y pueden qdar campos nulos.
```
perfect!

alangarciac commented on src/main/java/com/app/rest/controller/ItemController.java:
```diff
@@ -0,0 +1,82 @@
+        }
+    }
+    @PutMapping("/items/{id}")
+    public ResponseEntity<?> updateItem(@PathVariable("id") Long id, @RequestBody ItemDTO itemDTO) {
```
retornaría el item editado => ResponseEntity<ItemDTO>

cdpiazza replied:
Lo devuelve, el <?> es por lo mismo q mas arriba.

alangarciac replied:
oka

alangarciac commented on src/main/java/com/app/rest/model/dto/ItemDTO.java:
```diff
@@ -0,0 +1,94 @@
+        this.description = description;
+        this.deleted = deleted;
+    }
+    public ItemDTO(ItemDetail itemDetail) {
```
faltaria el `IllegalArgumentException` en el method signature

cdpiazza replied:
Pasa que eso lo hace el validate, no el constructor. Fijate q por comentario dice que quiza no sea ni necesario ese try-catch

alangarciac commented on src/main/java/com/app/rest/model/dto/ItemDTO.java:
```diff
@@ -0,0 +1,94 @@
+        this.description = itemDetail.getDescription();
+        this.deleted = itemDetail.isDeleted();
+        try {   //ACA NO SE SI ES NECESARIO ESTE TRY-CATCH, pasa q en el order vi que lo hiciste con OPTIONAL y sino tira una excepcion, y esto entiendo seria la manera sin usar esa manera java8 de programar.
+            this.type = ItemType.fromString(itemDetail.getType());
```
Acá hay un potencial `NullPointerException` si se elimina accidentalmente un tipo del `ItemType`. Fijate que en Order, justametne se llama a `Optional.ofNullable()` porque el metodo `fromString` puede retornar `null`.

cdpiazza replied:
Pero la asignacion de un Null no es un problema en si, despues el validate salta porq dice null no es un tipo valido y listo. No alcanza con eso y sacar el try-catch?

alangarciac replied:
Acá se esta instanciando un ItemDTO a partir de un ItemDetail. Sería el caso que se recupera un Item de la DB para usarlo internamente como DTO.

Si en algun momento se elimina o modifica un tipo de ItemType, puede que al momento de resolver el type, quede el type en `null`. Si luego de instanciar este itemDTO en alguna parte se llama a `itemDTO.getType()` tendríamos un NPE(NullPointerException). Habría que levantar una exception pateando en la creación, para no dejar "pasar" un ItemDTO con type null, sabiendo que se recupero internametne de la base.

cdpiazza replied:
Tecnicamente no te salta el NPE, itemDTO.getType te va a devolver null, porque puede ser null, pero se romperia al qrer usarlo...

En todo caso, supongamos q pasa lo que decis, levantas un tipo null de la bbdd, en el constructor levantas la excepcion, y se rompe, y nunca podes acomodar el tipo del objeto persistente, porq no lo podes recuperar, como lo arreglas? tenes q ir a tocar la BBDD directamente...

alangarciac replied:
En ese caso sí, a lo que apunta es a romper lo antes posible.

_se romperia al qrer usarlo_ ahí esta el problema, siempre es mejor que rompa rápido, en este caso cuando se está instanciando. A la larga, ahorra tiempos de debug para buscar el null y se detecta mas rápido. ItemDTO sabe como instanciarse y debería chillar cuando no puede completar un campo escencial, el type por ejemplo. Dejar que pase ese null, sería fallar en silent.

Fijate esto -> https://www.yegor256.com/2015/08/25/fail-fast.html

cdpiazza replied:
ok, entonces levanto una excepcion ahi y la manejo siempre q este instanciando un DTO a partir de un itemdetail?

cdpiazza replied:
como se resuelve lo de tener ir a tocar la BBDD para arreglarlo?

alangarciac replied:
si, es valido que el ctor de itemDTO a partir de ItemDetail tire una exception si no pudo inferir el type

alangarciac replied:
se resuelve yendo a la BBDD en ese caso, seria parte de una tarea que modifica tipos de items, constatar que los datos queden consistentes

cdpiazza replied:
oka, entonces listo.

alangarciac commented on src/main/java/com/app/rest/model/dto/ItemDTO.java:
```diff
@@ -0,0 +1,94 @@
+        this.deleted = itemDetail.isDeleted();
+        try {   //ACA NO SE SI ES NECESARIO ESTE TRY-CATCH, pasa q en el order vi que lo hiciste con OPTIONAL y sino tira una excepcion, y esto entiendo seria la manera sin usar esa manera java8 de programar.
+            this.type = ItemType.fromString(itemDetail.getType());
+        }catch (IllegalArgumentException ia) {
```
Quien estaría levantando la exception? No veo el **throw** de `IllegalArgumentException`

cdpiazza replied:
No se, lo puse por migrar del optional a esto, pense q el from string podria llegar a tirar una excepcion, pero no, solamente devuelve null, asique no es necesario.

alangarciac replied:
claro, se podría hacer que el fromString retorne esta exception, ó trabajarlo como un Option.ofNulleable(). Ambos son casos de uso válidos.

cdpiazza replied:
ok

alangarciac commented on src/main/java/com/app/rest/service/ItemService.java:
```diff
@@ -1,30 +1,82 @@
 
 @Service
 public class ItemService {
+    private static Logger LOGGER = LoggerFactory.getLogger(OrderService.class);
```
el `LOGGER` debería ser de `ItemService`

cdpiazza replied:
fue copy-paste

alangarciac commented on src/main/java/com/app/rest/service/ItemService.java:
```diff
@@ -1,30 +1,82 @@
-        return itemRepo.findAll();
+    public ItemDTO getItemById(Long id) {
+        try {
+            return new ItemDTO(itemRepo.findById(id).get());
```
El uso de los `Optional` suelen usarse con el `isPresent()` antes de llamar a un `.get()`

cdpiazza replied:
ok

alangarciac commented on src/main/java/com/app/rest/service/ItemService.java:
```diff
@@ -1,30 +1,82 @@
+    public ItemDTO getItemById(Long id) {
+        try {
+            return new ItemDTO(itemRepo.findById(id).get());
+        } catch (NoSuchElementException e) {
```
Retornaría una custom Exception, como `ItemNotFoundException()`

cdpiazza replied:
ok

alangarciac commented on src/main/java/com/app/rest/service/ItemService.java:
```diff
@@ -1,30 +1,82 @@
+    public List<ItemDTO> getItems()  {
+        List<ItemDTO> itemsDTO = new ArrayList<ItemDTO>();
+        try {
+            List<ItemDetail> list = itemRepo.findAll();
```
`List<ItemDetail> items` en lugar de > `List<ItemDetail> list`

cdpiazza replied:
ok

alangarciac commented on src/main/java/com/app/rest/service/ItemService.java:
```diff
@@ -1,30 +1,82 @@
+        List<ItemDTO> itemsDTO = new ArrayList<ItemDTO>();
+        try {
+            List<ItemDetail> list = itemRepo.findAll();
+            for (ItemDetail x : list) {
```
`itemDetail` en lugar de `x`

cdpiazza replied:
ok

alangarciac commented on src/main/java/com/app/rest/service/ItemService.java:
```diff
@@ -1,30 +1,82 @@
+                itemsDTO.add(new ItemDTO(x));
+                }
+            return  itemsDTO;
+        } catch (DataAccessException ie) {
```
Acá es donde no me cierra tener una interface como Repo. `ItemService` no debería handlear una Exception de `DataAccess` es responsabildiad del DAO/Repository handlear este tipo de Exceptions.

cdpiazza replied:
Eso es simplemente por como se usa hibernate, si hubiera un error de acceso a la BBDD, igualmente habria una exception, si queres de otro tipo, pero igualemten habria algo que handlear en el service.

alangarciac replied:
el Service tendría que enterarse que no se pudieron recuperar los Items. Por ejemplo (ItemException) o un IllegalStateException, es decir, `DataAccessException` es un error de acceso al dato, quien se encarga de buscar los datos deberá manejarlo y throwear _no lo pude recuperar_ el porqué, se loguea como error en este caso.

cdpiazza replied:
Esta bien, pero la interfaz provee esa manera de decir no lo pude recuperar, entonces lo enmascaro en el service y levanto un ItemException.

alangarciac commented on src/main/java/com/app/rest/service/ItemService.java:
```diff
@@ -1,30 +1,82 @@
+        }
+    }
+
+    public ItemDTO saveItem(ItemDTO itemDTO) { //Esta bien que no haya control de duplicados?
```
`//Esta bien que no haya control de duplicados?` > Debería hacerse el control de duplciados antes de llamar al save, verificando por ejemplo que el item no existe en la base. El criterio de `equal` podríamos definirlo como `name`, `type` y alguna otra cosa que consideres como necesaria para el equal

cdpiazza replied:
En el service ?

alangarciac replied:
si, el service sería el que determina si esta o no duplicado y como proceder, ya sea por error o por success

cdpiazza replied:
oki

alangarciac commented on src/main/java/com/app/rest/controller/ItemController.java:
```diff
@@ -0,0 +1,82 @@
+            return ResponseEntity.ok(itemDTO);
+        } catch (IllegalArgumentException is) {
+            return ResponseEntity.badRequest().body(is.getMessage());
+        } catch (DataAccessException ie) {
```
El controller no debería handlear un `DataAccess` propio del `@Repository`

cdpiazza replied:
Defino una exception nueva para cuando hay un error de conexion a la bbdd y la enmascaro? o como le aviso al controller q se rompio la bbdd?

alangarciac commented on src/main/java/com/app/rest/service/ItemService.java:
```diff
@@ -1,30 +1,82 @@
+            return new ItemDTO(itemRepo.save(itemDetail));
+        } catch (DataAccessException ie) {
+            LOGGER.error("Error trying to save item {}. Caused by[{}]", itemDetail, ie.getMessage());
+            throw ie;
```
podría enmascararse en un `ItemException` para no propagar el `DataAccess`

cdpiazza replied:
ok, lo mismo preguntaba mas arriba, hacemos asi.

alangarciac commented on src/main/java/com/app/rest/service/ItemService.java:
```diff
@@ -1,30 +1,82 @@
+            itemDetail.setName(itemDTO.getName());
+            itemDetail.setDescription(itemDTO.getDescription());
+            itemDetail.setType(itemDTO.getType());
+            itemDetail.setDeleted(itemDTO.isDeleted());
```
el `setDeleted` solo debería llamarse en el `delete`

cdpiazza replied:
pero es un update, nunca pueden updatear ese campo del objeto si no es en el delete?

alangarciac replied:
mmm no, para eso es el `delete`. el campo justamente se updatea con el /delete

cdpiazza replied:
oki!

alangarciac commented on src/main/java/com/app/rest/service/ItemService.java:
```diff
@@ -1,30 +1,82 @@
+            itemDetail.setDescription(itemDTO.getDescription());
+            itemDetail.setType(itemDTO.getType());
+            itemDetail.setDeleted(itemDTO.isDeleted());
+            return new ItemDTO(itemRepo.save(itemDetail));
```
No es necesario, pero solo a modo comentario, se suele poner un attribute updated_at o last_modified como date para setear la fecha en la que se editó por ultima vez

cdpiazza replied:
Queres q edite el modelo para contemplarlo?

alangarciac replied:
como quieras, simplemente comenté lo que se suele hacer

cdpiazza replied:
ok, lo hago, voy a tener q tocar el .sql

alangarciac commented on src/test/java/com/app/rest/test/service/ItemTest.java:
```diff
@@ -0,0 +1,59 @@
+public class ItemTest {
+
+    @Autowired
+    private ItemService itemService;
```
funcó el `mvn test` con este @Autowired?

cdpiazza replied:
nunca lo probe a mvn test, anda corriendolo a mano como test dandole run.

alangarciac commented on src/test/java/com/app/rest/test/service/ItemTest.java:
```diff
@@ -0,0 +1,59 @@
+        try {
+            itemService.getItemById(8L);
+        } catch (NoSuchElementException nf) {
+            System.out.println("Id 8 not found, which is expected");
```
con JUnit se pueden hacer asserts si levanta una Exception y es esperada -> https://stackoverflow.com/questions/156503/how-do-you-assert-that-a-certain-exception-is-thrown-in-junit-4-tests/2935935#2935935

cdpiazza replied:
Si pero esto no es junit, ni siquiera son test de unidad, son cosas q hice para ir probando q retornaba y que venia, sin nteer q andar usando el controller, no son TESTS SERIOS. No se hacer tests serios.

alangarciac replied:
jajajjaa bueno, vamos a ir incorporándolos, suma ir metiendo tests

alangarciac commented on src/main/java/com/app/rest/controller/ItemController.java:
```diff
@@ -0,0 +1,82 @@
+            return ResponseEntity.notFound().build();
+        }
+    }
+    @GetMapping("/items")
```
faltaría sumar el `/items/{type}` que retorne los items de ese `type`

cdpiazza replied:
Eso es una feature para mas adelante.

alangarciac reviewed (changes requested):
Impecable, algunos detalles de names y referencias.

faltaría sumar el /items/{type} que retorne los items de ese type

Commit:
Cambios a ser confirmados:
	modificado:     src/main/java/com/app/rest/controller/ItemController.java
	nuevo archivo:  src/main/java/com/app/rest/exception/ItemException.java
	nuevo archivo:  src/main/java/com/app/rest/exception/ItemNotFoundException.java
	modificado:     src/main/java/com/app/rest/service/ItemService.java
	modificado:     src/test/java/com/app/rest/test/service/ItemTest.java

Commit:
Cambios a ser confirmados:
	modificado:     src/main/java/com/app/rest/service/ItemService.java

Commit:
Cambios a ser confirmados:
	modificado:     src/main/java/com/app/rest/controller/ItemController.java
	nuevo archivo:  src/main/java/com/app/rest/exception/ItemTypeException.java
	nuevo archivo:  src/main/java/com/app/rest/exception/ItemValidateException.java
	modificado:     src/main/java/com/app/rest/model/dto/ItemDTO.java
	modificado:     src/main/java/com/app/rest/model/dto/ItemType.java
	modificado:     src/main/java/com/app/rest/model/persistence/ItemDetail.java
	modificado:     src/main/java/com/app/rest/repository/ItemRepo.java
	modificado:     src/main/java/com/app/rest/service/ItemService.java
	modificado:     src/main/resources/import.sql
	modificado:     src/test/java/com/app/rest/test/service/ItemTest.java

cdpiazza reviewed (commented):
Cambios solicitados realizados.
Nuevo commit a espera de ser revisado

alangarciac commented on src/main/java/com/app/rest/model/dto/ItemDTO.java:
```diff
@@ -0,0 +1,104 @@
+    }
+
+    public void validate () throws ItemValidateException{
+        if (this.name == null || this.description == null){ //type boolean cant be null, when null->false, which is ok here.
```
acá podés usar `StringUtils.hasText(name)`, chequea null y string vacío. Internamente hace esto:

```
    public static boolean hasText(@Nullable String str) {
        return str != null && !str.isEmpty() && containsText(str);
    }
```

cdpiazza replied:
okidoki

alangarciac commented on src/main/java/com/app/rest/model/dto/ItemType.java:
```diff
@@ -0,0 +1,31 @@
+                return itemType;
+            }
+        }
+        throw new ItemTypeException("Item Type not valid, types admitted:" + Arrays.toString(values()));
```
perfect!

alangarciac commented on src/main/java/com/app/rest/service/ItemService.java:
```diff
@@ -1,30 +1,132 @@
 
-    public List<ItemDetail> getItems() throws InterruptedException {
-        return itemRepo.findAll();
+    static Specification<ItemDetail> hasName(String name) {
```
`private` ?

cdpiazza replied:
Si , me lo olvide :dagger:

alangarciac commented on src/main/java/com/app/rest/service/ItemService.java:
```diff
@@ -1,30 +1,132 @@
     }
-    public Optional<ItemDetail> getItem(Long id) throws InterruptedException {
-        return itemRepo.findById(id);
+    static Specification<ItemDetail> descriptionContains(String description) {
```
`private` ?

cdpiazza replied:
si! me lo olvide :D

alangarciac reviewed (dismissed):
cambiaría la visibilidad de los metodos del Service a `private` si se usan internamente

Commit:
Cambios a ser confirmados:
	modificado:     src/main/java/com/app/rest/model/dto/ItemDTO.java
	modificado:     src/main/java/com/app/rest/service/ItemService.java

alangarciac reviewed (approved):
perfect
